src/model/UserDataSource.ts

```diff
@@ -0,0 +1,200 @@
+import { MongoDataSource } from 'apollo-datasource-mongodb'
+import { MUUID } from 'uuid-mongodb'
+import differenceInDays from 'date-fns/differenceInDays/index.js'
+
+import { getUserModel } from '../db/index.js'
+import { User, UpdateProfileGQLInput, UsernameInfo, GetUsernameReturn } from '../db/UserTypes.js'
+import { trimToNull } from '../utils/sanitize.js'
+
+const USERNAME_UPDATE_WAITING_IN_DAYS = 14
+
+interface UsernameQueryReturnType {
+  username?: string
+  createdAt?: string
+  userUuid: string
+}
+export default class UserDataSource extends MongoDataSource<User> {
+  userModel = getUserModel()
+
+  /**
+   * Update user profile.  Create a new user object if not defined.
+   * @param userUuid userUuid to update/add
+   * @param input profile params
+   * @returns true if successful
+   */
+  async createOrUpdateUserProfile (updater: MUUID, input: UpdateProfileGQLInput): Promise<boolean> {
+    const {
+      username: _username,
+      displayName: _displayName,
+      bio: _bio,
+      website: _website,
+      email,
+      emailVerified,
+      _id
+    } = input
+
+    if (Object.keys(input).length === 0) {
+      throw new Error('Nothing to update. Must provide at least one field.')
+    }
+
+    const username = trimToNull(_username)
+    const displayName = trimToNull(_displayName)
+    const bio = trimToNull(_bio)
+    const website = trimToNull(_website)
+
+    if (username == null && displayName == null && bio == null && website == null) {
+      throw new Error('Nothing to update. Must provide at least one field.')
+    }
+
+    if (username != null && !isValidUsername(username)) {
+      throw new Error('Invalid username format.')
+    }
+
+    if (website != null && !isValidUrl(website)) {
+      throw new Error('Invalid website address.')
+    }
+
+    const rs = await this.userModel.findOne({ _id })
+
+    const isNew = rs == null
+
+    if (isNew) {
+      if (email == null) {
+        throw new Error('Email is required when creating a new profile')
+      }
+      let usernameInfo: UsernameInfo | null = null
```

Comment:
I'm also wondering -- username seems so central. Why don't we want to force creation of a username right off the bat? Furthermore, we could force a user-inputted rather than auto-generated username to avoid issues like this: https://github.com/OpenBeta/open-tacos/issues/750. A bit more friction, but probably makes the site seem more human when there are human names everywhere as opposed to robot-like names.

Reply:
As discussed in the other comment, new open-tacos users will be required to create a username.  Allowing usernameInfo object to be nullable in the backend will let us create new user accounts programmatically (a future use case).